packages/web3-common/src/formatter.ts

```diff
@@ -0,0 +1,254 @@
+import {
+	Bytes,
+	bytesToBuffer,
+	bytesToHex,
+	HexString,
+	mergeDeep,
+	Numbers,
+	numberToHex,
+	toBigInt,
+} from 'web3-utils';
+import { isObject, JsonSchema, utils, ValidationSchemaInput } from 'web3-validator';
+import { FormatterError } from './errors';
+
+const { parseBaseType } = utils;
+
+export enum FMT_NUMBER {
+	NUMBER = 'NUMBER_NUMBER',
+	HEX = 'NUMBER_HEX',
+	STR = 'NUMBER_STR',
+	BIGINT = 'NUMBER_BIGINT',
+}
+
+export type NumberTypes = {
+	[FMT_NUMBER.NUMBER]: number;
+	[FMT_NUMBER.HEX]: HexString;
+	[FMT_NUMBER.STR]: string;
+	[FMT_NUMBER.BIGINT]: bigint;
+};
+
+export enum FMT_BYTES {
+	HEX = 'BYTES_HEX',
+	BUFFER = 'BYTES_BUFFER',
+	UINT8ARRAY = 'BYTES_UINT8ARRAY',
+}
+
+export type ByteTypes = {
+	[FMT_BYTES.HEX]: HexString;
+	[FMT_BYTES.BUFFER]: Buffer;
+	[FMT_BYTES.UINT8ARRAY]: Uint8Array;
+};
+
+export type DataFormat = {
+	readonly number: FMT_NUMBER;
+	readonly bytes: FMT_BYTES;
+};
+
+export const DEFAULT_RETURN_FORMAT = { number: FMT_NUMBER.HEX, bytes: FMT_BYTES.HEX } as const;
+
+// Added `undefined` to cover optional type
+export type FormatType<T, F extends DataFormat> = number extends Extract<T, Numbers>
+	? NumberTypes[F['number']] | Exclude<T, Numbers>
+	: Buffer extends Extract<T, Bytes>
+	? ByteTypes[F['bytes']] | Exclude<T, Bytes>
+	: T extends object | undefined
+	? {
+			[P in keyof T]: FormatType<T[P], F>;
+	  }
+	: T;
+
+const findSchemaByDataPath = (schema: JsonSchema, dataPath: string[]): JsonSchema | undefined => {
+	let result: JsonSchema = { ...schema } as JsonSchema;
+
+	for (const dataPart of dataPath) {
+		if (!result.properties && !result.items) {
+			return undefined;
+		}
+
+		if (result.properties) {
+			result = (result.properties as Record<string, JsonSchema>)[dataPart];
+		} else if (result.items && (result.items as JsonSchema).properties) {
+			const node = (result.items as JsonSchema).properties as Record<string, JsonSchema>;
+
+			if (!node) {
+				return undefined;
+			}
+
+			result = node[dataPart];
+		} else if (result.items && isObject(result.items)) {
+			result = result.items;
+		} else if (result.items && Array.isArray(result.items)) {
+			result = (result.items as JsonSchema[])[parseInt(dataPart, 10)];
+		}
+	}
+
+	return result;
+};
+
+export const convertScalarValue = (value: unknown, ethType: string, format: DataFormat) => {
```

Comment:
Do you think this should throw an error if both `if`s are falsy? Something like a `UnsupportedEthTypeError`?

Reply:
We have to return original value e.g. If someone uses `eth: 'string'`, then we dont' need to format and we can simply return the original value.